Limit RSA-OAEP padding to SHA-1 and SHA-2 algorithms in supported openssl

collisdigital opened the pull request:
Raising this as WIP, want your thoughts on this approach. I've updated the openssl backend to limit the hashes allowed with RSA-OAEP depending on the OpenSSL version to those verified with test vectors. 

Doing this made some of the code in rsa.py a duplicate of the backend.py so I've removed the duplicate checks, however it does mean the errors are less verbose when an invalid combination is given. Advantage being only one set of checks need to be maintained and two areas of code don't need to be synchronised manually when changing.

What do you think?

reaperhulk commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -578,9 +578,25 @@ def _pem_password_cb(self, password):
+        # with OpenSSL >= 1.0.2.
+        if backend._lib.OpenSSL_version_num() >= 0x10002001:
+            _oaep_supported_padding_algorithms = (
+                hashes.SHA224, hashes.SHA256,
```
SHA1 still needs to be in this list :)

collisdigital replied:
Yeah I noticed this too, although it does pass the test with SHA-1 because of the way the logic is written but I don't think it should rely on that. However, it raised another question to me, we only have test vectors that cover the SHA-2 combinations not SHA-1 with SHA-2 in the same OAEP. Do we want to add vectors to test SHA-1+SHA-2 combinations or put in a restriction to mark this combination as unsupported?

reaperhulk commented:
I'll have to think about the change to the errors. It is simpler, but for users on older OpenSSL (esp. < 1.0.2) it may be insufficiently descriptive to avoid users filing bugs. Thanks for working on this!

collisdigital commented:
I've added SHA-1 explicitly to the supported list and also added a check for SHA-1 being used in combination with SHA-2 which doesn't have test vectors (yet?). Keeping the code clean with these kinds of logic checks is tricky.

collisdigital commented:
I've put the checks back into rsa.py as I agree with you that it is likely to cause confusion without them.

reaperhulk commented:
Good point on the missing SHA-1 combination vectors. We should generate those as well.

reaperhulk commented:
I've updated the vector PR with the missing combination vectors and updated the generation/verification code as well.

collisdigital commented:
Nice one, I'll remove the limitation check now

collisdigital commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -578,7 +578,40 @@ def _pem_password_cb(self, password):
+        return _oaep_supported_padding_algorithms
+
+    def _oaep_mgf1_hash_supported(self, algorithm):
+        if self._lib.Cryptography_HAS_MGF1_MD:
```
@reaperhulk It's not clear to me what the implications of this check are, _oaep_hash_supported will return it supports SHA-1 and SHA-2 in the current code but _oaep_mgf1_hash_supported will only return it supports SHA-1 when this HAS_MGF1_MD is false, is that the right behaviour?

Is HAS_MGF1_MD linked tightly to the OpenSSL version? And if so is it actually covered by _get_oaep_supported_padding_algorithms check for version >= 0x10002001 and so this HAS_MGF1_MD isn't needed and can be removed etc?

reaperhulk replied:
Yep, you can remove this check.

collisdigital commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -578,7 +578,40 @@ def _pem_password_cb(self, password):
+
+    def _pss_mgf1_hash_supported(self, algorithm):
         if self._lib.Cryptography_HAS_MGF1_MD:
             return self.hash_supported(algorithm)
```
I have split out the PSS check as I wasn't sure that these changes should also apply the PSS? Although the existing code returns True for any supported hash with PSS, is that correct? If not I think that's a separate PR...

reaperhulk replied:
Yeah correct. The existing code is perhaps too lax (although that's arguably a part of our public API now and it's likely no one has used anything other than SHA1/SHA2).

collisdigital commented:
Removed check, only have 6% battery on MacBook and left charger at work, so hope that's the last change ;)

reaperhulk commented:
Going to go ahead and merge this in and we can continue review on the main PR as needed.
